server/writer/writer.go

```diff
@@ -49,21 +49,23 @@ func ErrorString(w http.ResponseWriter, status int, code string, err error) {
 
 // Error writes a response with specified status and error response.
 func Error(w http.ResponseWriter, status int, err *types.ErrorV1) {
-	headers := w.Header()
-	headers.Add("Content-Type", "application/json")
+	w.Header().Set("Content-Type", "application/json")
 	w.WriteHeader(status)
 	_, _ = w.Write(append(err.Bytes(), byte('\n')))
 }
 
 // JSON writes a response with the specified status code and object. The object
 // will be JSON serialized.
+// Deprecated: This method is problematic when using a non-200 status `code`: if
+// encoding the payload fails, it'll print "superfluous call to WriteHeader()"
+// logs.
 func JSON(w http.ResponseWriter, code int, v interface{}, pretty bool) {
```

Comment:
There is only a single caller to this method, and it's the health check logging an error. The payloads involved with that are so simple that this problem shouldn't really apply -- but we could refactor this nonetheless.